org.eclipse.lsp4e/src/org/eclipse/lsp4e/format/IFormatRegionsProvider.java

```diff
@@ -0,0 +1,32 @@
+/*******************************************************************************
+ * Copyright (c) 2023 Contributors to the Eclipse Foundation.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   See git history
+ *******************************************************************************/
+package org.eclipse.lsp4e.format;
+
+/**
+ * Can be implemented by clients as OSGi service
+ * to provide editor specific formatting regions for the format-on-save feature.
+ * The OSGi component service has to implement the {@code serverDefinitionId} property.
+ * The value must be the {@code server id} of the corresponding {@code languageServer} extension point.
+ * This service will then be used for documents who are connected to this language server.
+ * <p>Example:
+ * <pre>{@code
+ * @Component(property = { "serverDefinitionId:String=org.eclipse.cdt.lsp.server" })
+ * public interface IFormatRegionsProvider {
+ *  	IRegion[] getFormattingRegions(IDocument document){
+ *  		//formats whole document:
+ * 			new IRegion[] { new Region(0, document.getLength()) };
+ * }
+ * }</pre>
+ */
+public interface IFormatRegionsProvider extends IFormatRegions {
+
```

Comment:
Inheritance quickly becomes a burden. I would instead suggest here

```java
public interface IFormatRegionsStrategyProvider {
    /**
     * @return an IFormatRegionsComputer suitable for this document,
     * or <code>null</code> if the current provider cannot equip the document
     * for formatting regions.
     */
    public IFormatRegionsComputer getFormatRegionsComputer(Document document);
}
```
and FormatAllLins and FormatEditedLines would implement IFormatRegionsComputer.

So consumers could just implement eg
```java
public class FormatEditedLinesOnSaveForMyJSXFiles implements IFormatRegionsStrategyProvider {
   public IFormatRegionsComputer getFormatRegionsComputer(Document document) {
     if (!isJSXFile(document)) {
       return null;
     }
     return new FormatEditedLines();
  }
}
```

Reply:
The proposal above will lead to this call in `DocumentContentSynchronizer`:
```java
return formatRegionsStrategyProvider.getFormatRegionsComputer(document).getFormattingRegions(document);
```
This will fail if `getFormatRegionsComputer(document)` returns `null`, So a check for `null` must be applied.

The only reason for:
```java
public interface IFormatRegionsProvider extends IFormatRegions 
```
is to distinguish between the provider (`IFormatRegionsProvider`) and the computer (`IFormatRegions`). The provider  should be implemented as OSGI service . It contains the computer implementation (`IFormatRegions`). In that case, the call is even simpler:
```java
return formatRegionsProvider.getFormattingRegions(document);
```